### PR TITLE
allow main source file to have not only .cc extension

### LINKED ISSUE #980: Support for .cpp files
poac build fails if main file is not called main.cc. I saw another issue about the sane problem, I a using the latest version of poac and cannot build a project where main file has extension .cpp. Is it intended to use only main.cc filename or I can change it somehow?

### PR DESCRIPTION
**Pull request by yaito3014**
fixes #980

### CODE REVIEW

**Comment by zkkv** [comment]
Does this allow any file that has `main` before period? Perhaps, it better to allow a finite set of extensions?

**Comment by yaito3014** [comment]
> Does this allow any file that has `main` before period? Perhaps, it better to allow a finite set of extensions?

`listSourceFilePaths` function already lists only source files with an extension in `SOURCE_FILE_EXTS`, so main source file must have one of those extensions.

**Comment by ken-matsui** [comment]
Thank you for working on this!

> > Does this allow any file that has `main` before period? Perhaps, it better to allow a finite set of extensions?
> 
> `listSourceFilePaths` function already lists only source files with an extension in `SOURCE_FILE_EXTS`, so main source file must have one of those extensions.

`listSourceFilePaths` finds source files recursively, so `src/foo/main.cpp` could be used for example but this should be treated as a normal source file, although this is awkward (for better usability, we can consider emitting a warning as it's likely a human error). Also, if there are multiple main source files, which file will be chosen is ambiguous. So, I think the best rules will be:

1. There should be only one main source file., if not, emit an error (multiple main files).
2. The main file should be directly in the `src` directory, if not, emit an error (not found).

Eventually, we can also implement the `src/bin` binary target feature and those rules wouldn't interfere with this feature.

I would love to hear your thoughts!

File: src/BuildConfig.cc [hunk 1]
```diff
@@ -771,9 +772,27 @@ configureBuild(BuildConfig& config, const bool isDebug) {
...
+
+  // find main source file
+  fs::path mainSource;
+  for (const auto& entry : fs::directory_iterator("src")) {
+    fs::path path = entry.path();
+    if (!SOURCE_FILE_EXTS.contains(path.extension())) {
+      continue;
+    }
+    if (const auto& s = path.filename().string();
+        s.substr(0, s.find_last_of('.')) != "main") {
```
Comment (ken-matsui) [inline src/BuildConfig.cc:784]: Let's create a small function called `isMainSource` that checks if a string is a main source.

File: src/BuildConfig.cc [hunk 2]
```diff
@@ -793,9 +812,18 @@ configureBuild(BuildConfig& config, const bool isDebug) {
   std::vector<fs::path> sourceFilePaths = listSourceFilePaths("src");
   std::string srcs;
   for (fs::path& sourceFilePath : sourceFilePaths) {
+    if (const auto& s = sourceFilePath.filename().string();
+        sourceFilePath != mainSource
+        && s.substr(0, s.find_last_of('.')) == "main") {
+      logger::warn(
+          "source file named main was found not directly in src directory"
```
Comment (ken-matsui) [inline src/BuildConfig.cc:819]: This warning could be emitted multiple times, so specifying the exact path would make more sense, e.g.,:

```suggestion
          fmt::format("`{}` is named `main` but is not located directly in the `src/` directory. This file will not be treated as the program's entry point.", sourceFilePath.string())
```

Feel free to modify this if needed.

**Review by ken-matsui — COMMENTED** [review]
Just small things, and LGTM afterward!  Thank you!

**Review by ken-matsui — APPROVED** [review]
Thank you!

**Pull request merged**